Support Bind to address for middleman

maxmeyer opened the pull request:
Hi there,

some time ago @bhollis implemented 08dee580aa2c99205b8d702f54e9d9087048d612. Unfortunately this broke https://github.com/fedux-org/middleman-presentation because it used the `-h` of `middleman server`. This PR is a follow up of #1516.

The `middleman-presentation`-gem is used to serve `reveal.js`-presentations via `middleman server`  - besides some other things. As I'm connected to wifi during some talks I tend to bind the server to `localhost` to prevent it to be public available.

I would like to appreciate to make it possible for users to explicitly define a bind address again. This PR implements this again + a name resolution for the given bind address.

BTW:
Is `middlmean` a project using http://semver.org? Can I be sure that breaking changes only happen on major or a least on minor versions (minor versions during transition to v4).

Ping @tdreyno as well.

tdreyno commented:
@maxmeyer We are "semver-ish". Basically, I'm fine adding new features in patch releases. A functional change, that may have side-effects would bump minor (see the i18n link_to changes). Complete feature removal or API change in the major level.

tdreyno commented:
I'll let @Arcovion and @bhollis decide on this PR since they're way more familiar with the bind stuff

Arcovion commented:
The default `bind_address` needs to be `0.0.0.0`, else this breaks `localhost:4567` on Windows.

maxmeyer commented:
That's not clear to me. Does `localhost` not work, if `middleman` listens on 127.0.0.1?

Arcovion commented:
The default here is binding to `public_ip`, which is not `127.0.0.1`.

So `bundle exec middleman` &rarr; `PublicIP:4567` works, `Hostname:4567` works but `Localhost:4567` doesn't work so this isn't backwards-compatible.

Using `bundle exec middleman -b 127.0.0.1` works for `localhost`, but then the `PublicIP:4567` doesn't work.

But `bundle exec middleman -b 0.0.0.0` works with everything, so this should remain the default and you can use `-b` to limit it to `localhost` or whatever.

maxmeyer commented:
Please forget about my last comment. I think I understand now. Will fix this.

Arcovion commented:
Before this commit the `bind_address` was unset, which defaults to `nil`. 
That means ["0.0.0.0 or ::"](http://docs.ruby-lang.org/ja/2.0.0/method/WEBrick=3a=3aConfig/c/HTTP.html), so set it to `nil` by default and not `0.0.0.0`, to support IPv6.

bhollis commented on middleman-core/lib/middleman-core/preview_server.rb:
```diff
@@ -24,7 +25,7 @@ def start(opts={})
 
         logger.debug %(== The Middleman is running in "#{environment}" environment)
-        logger.info "== The Middleman is standing watch at #{uri} (#{uri(public_ip)})"
+        logger.info "== The Middleman is standing watch at #{uri} (#{uri(bind_address)})"
```
Won't this pass `nil` into `uri` by default?

maxmeyer replied:
Yes. I think so.

bhollis commented on middleman-core/lib/middleman-core/preview_server.rb:
```diff
@@ -110,24 +111,29 @@ def new_app
         end
 
+        my_ip       = opts[:bind_address].presence || public_ip
+        my_hostname = opts[:host].presence || resolve_ip(my_ip) || Socket.gethostname.tr(' ', '+')
```
Previously we had people complaining about having to do a reverse DNS lookup at any point (because they were on the train or whatever). What does `resolve_ip` get us that `Socket.gethostname` doesn't?

maxmeyer replied:
The problem is this, `middleman` tells you it's listening on `http://host.in.mydomain.org:4567/`

```
% middleman
== Could not find a Middleman project config.rb
== Treating directory as a static site to be served
== The Middleman is loading
== The Middleman is standing watch at http://host.in.mydomain.org:4567/ (http://192.168.1.2:4567/)
== Inspect your site configuration at http://host.in.mydomain.org:4567/__middleman
```

Looking at `netstat` you get this:

```
netstat -nlpt
tcp        0      0 0.0.0.0:4567            0.0.0.0:*               LISTEN      7222/ruby    
```

But doing a reverse lookup in a SOHO-environment with a simple dns forwarder fails for the local ip address:

```
host 192.168.1.2
Host 2.1.168.192.in-addr.arpa. not found: 3(NXDOMAIN)
```

and a normal lookup does not work either

```
% host $(hostname)
Host host.in.mydomain.org not found: 3(NXDOMAIN)
```

> reverse DNS lookup at any point (because they were on the train or whatever)

I think that was me using my coporate account. The resolve_ip thing is only used if you do not pass a hostname. I will investigate a little bit further and will try to build a solution which works for a coporate and for SOHO-environments. Ok?

sandstrom commented:
Partly related, I think it's important that middleman listens to localhost by default (instead of 0.0.0.0 i.e. every machine on the network). Listening on 0.0.0.0 is a security issue and not only a theoretical one — people often work in coffeeshops, on wifi-airplanes etc.

Rails worked through a similar issue a while back:
https://github.com/rack/rack/commit/28b014484a8ac0bbb388e7eaeeef159598ec64fc

maxmeyer commented:
I added the first tests and some rough ideas how to build the additional classes. I think there are still some missing tests, e.g. for IPv6. Any feedback is welcome. :smile:

Arcovion commented on middleman-core/lib/middleman-core/server_listener.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+
+    def find_network_interfaces
+      Socket.ip_address_list.find(-> { ['127.0.0.1'] }) { |ai| ai.ipv4? && !ai.ipv4_loopback? }.map { |ai| ai.ip_address }
```
`find`/`detect` will only return the first item, also the lambda is returning an array which won't respond to `ip_address`.

maxmeyer replied:
I should have mentioned, that the code is very very broken and only the tests should be reviewed. Sorry for being that unprecise.

Arcovion commented on middleman-core/lib/middleman-core/server_listener.rb:
```diff
@@ -0,0 +1,72 @@
+  class ServerListener
+    private
+
+    attr_reader :https
```
No need to have this, just use `@https` in the code. Also you could have avoided the `private`/`public` dance by calling `private :https`.

maxmeyer replied:
> No need to have this, just use @https in the code.

:smile: I know, but I follow a tip from Sandy Metz not to use instance variables, but use a method instead. But I think this is just a matter of style. Although I had a view uses cases where the tip proved to be useful. If you want me to use instance variables, I'm happy to change this.

Arcovion commented:
A config option for `bind` should be exposed in v4 to go with `host` and `port`: https://github.com/middleman/middleman/commit/a71589becde156b87968d63a5087e16c5ea5c5d6

Arcovion commented:
@sandstrom I am okay with that, since it's going to be easy to configure after this PR is merged.
As long as we add documentation that it needs to be changed in order to to test sites on multiple devices with wifi etc.
Ping @bhollis

maxmeyer commented:
FYI: I'm working on some code in [`aruba`](https://github.com/cucumber/aruba/pull/257/) which will hopefully make testing `middleman server` easier. I'm not aware of any tests for the server-command of middleman and would like to add some more tests before changing that command again.

The idea is to provide a feature/scenario like

```cucumber
Given I run `middleman server` interactively
and I stop it when the output contains:
"""
== Middleman
"""
```

Would such a step a helpful addition from your point of view?

tdreyno commented:
@maxmeyer Can you rebase?

maxmeyer commented:
No problem. Will do this in the next couple of days. For now I'm a bit short of time.

tdreyno commented:
Cool. Thanks. Going to push `3.4.0` when this is stable.

maxmeyer commented:
Rebased, but still not finished. WIP.

tdreyno commented:
Would really like to get `v3.4.0` out the door. Is the github version, with the initial attempts at this feature, currently broken?

maxmeyer commented:
Yes.

tdreyno commented:
@maxmeyer I would appreciate it. I'm going on holiday for 2 weeks starting Friday. Was hoping to get this out the door by then.

maxmeyer commented:
Ok. I will do my best :smile:

tdreyno commented:
Tokaido has this: https://github.com/tokaido/tokaido-dns

maxmeyer commented:
Thanks, but no need for this. `ruby-core` has everything I needed. I wrapped it in a [single class](https://github.com/maxmeyer/middleman/blob/feature/listener/middleman-core/lib/middleman-core/dns_resolver.rb).

maxmeyer commented:
@tdreyno 
Would be unfair to block the release too long and being not finshed then. I think it's better to merge this PR for 3.5 and not for 3.4. 

Merging before holidays has some risk ;-) And this is change which should be reviewed carefully: The name resolution for some error cases gives me some headaches actually. I think I need some more time to wrap my head around this one.

tdreyno commented:
I agree, but then I need to back out the breaking change from before... which I'm having a hard time isolating.

dg-ratiodata commented:
Mmh... Which breaking change you mean?

maxmeyer commented:
Ok. Will then try harder. But I'm not sure if it will work.

maxmeyer commented on middleman-core/features/cli/preview_server.feature:
```diff
@@ -0,0 +1,117 @@
+Feature: Run the preview server
```
@tdreyno First version of feature tests. Mostly testing happy path.

maxmeyer commented on middleman-core/lib/middleman-core/application.rb:
```diff
@@ -75,7 +75,11 @@ def self.root_path
     # Which server name should be used
     # @return [NilClass, String]
-    config.define_setting :host, nil, 'The preview host name'
+    config.define_setting :server_name, nil, 'The server name of preview server'
```
I renamed host to server_name. I think this is a breaking change though the old setting was more or less usesless from my point of view.

maxmeyer commented on middleman-core/lib/middleman-core/dns_resolver.rb:
```diff
@@ -0,0 +1,91 @@
+require 'resolv'
+require 'active_support/core_ext/object/blank'
```
This one is only for name resolution. The main things here:

* Local Resolver is used from ruby-core -> /etc/hosts on linux and some other path on windows
* this uses HOSTSRC variable to find a local hosts file -> introduced for local testing, but might worth to mention in the docs
* Local Resolver is used first -> then Network Resolver, the network resolver is only used if the local resolver fails.

maxmeyer commented on middleman-core/lib/middleman-core/preview_server.rb:
```diff
@@ -19,13 +21,23 @@ def https?
 
         logger.debug %(== The Middleman is running in "#{environment}" environment)
-        logger.info "== The Middleman is standing watch at #{uri} (#{uri(public_ip)})"
-        logger.info "== Inspect your site configuration at #{uri + '__middleman'}"
```
* There are listeners, that's where `netstat -nlplt` (gnu netstat) gives you listening services.
* Normal Urls to open in Browser
* Config urls to open in Browser

maxmeyer commented on middleman-core/lib/middleman-core/preview_server/information.rb:
```diff
@@ -0,0 +1,155 @@
+require 'active_support/core_ext/object/blank'
```
This one wraps how the information is retrieved. This gaves me the most headaches. Each class is for a specific use case. The validation checks are still missing. I will add them later/tomorrow.

maxmeyer commented on middleman-core/lib/middleman-core/preview_server/network_interface_inventory.rb:
```diff
@@ -0,0 +1,9 @@
+module Middleman
```
This one knows about local interfaces (of the developer machine).

maxmeyer commented on middleman-core/lib/middleman-core/preview_server/server_information.rb:
```diff
@@ -0,0 +1,135 @@
+require 'middleman-core/preview_server/validator'
+require 'middleman-core/preview_server/information'
+require 'middleman-core/preview_server/network_interface_inventory'
+
```
Here's where most of the magic happens.

* It asks for name resolution
* It asks for validation (is missing for now)

It needs some cleanup though.

maxmeyer commented on middleman-core/lib/middleman-core/preview_server/server_url.rb:
```diff
@@ -0,0 +1,31 @@
+module Middleman
```
Url builder... Nothing more.

maxmeyer commented on middleman-core/lib/middleman-core/preview_server/validator.rb:
```diff
@@ -0,0 +1,55 @@
+module Middleman
```
Needs to be implemented.

maxmeyer commented on middleman-core/lib/middleman-core/step_definitions/commandline_steps.rb:
```diff
@@ -0,0 +1,15 @@
+When /^I stop middleman if the output contains:$/ do |expected|
```
This contains some code of an upcomming `aruba`-version. I will let you know when/how it can be safely removed. I still working on that bits over at cucumber/aruba.

maxmeyer commented on middleman-core/spec/dns_resolver_spec.rb:
```diff
@@ -0,0 +1,178 @@
+require 'spec_helper'
+require 'middleman-core/dns_resolver'
```
Needs to be repaired...

maxmeyer commented on middleman-core/spec/spec_helper.rb:
```diff
@@ -3,3 +3,24 @@
 
 require 'coveralls'
 Coveralls.wear!
+
```
Use some stuff of rspec 3. :smile:

maxmeyer commented:
@tdreyno I pushed a very very first working version to that branch. It needs a lot of cleanup and needs to cover more edge cases and the validations are still missing. But it might be worth that you and/or @Arcovion and/or @bhollis give it a go. But please keep in mind it's not finished yet.

tdreyno commented:
Fascinating and complicated. Thanks for taking the lead on this!

maxmeyer commented:
>  and complicated. 

Yes, unfortunately.

dg-ratiodata commented on middleman-core/lib/middleman-core/preview_server.rb:
```diff
@@ -19,13 +21,27 @@ def https?
         @options = opts
+        @server_information = ServerInformation.new
+
+        the_app = new_app
```
This evaluates the middleman configuration so I needed to move it to a variable and place it before the warning.

dg-ratiodata commented on middleman-core/lib/middleman-core/preview_server/checks.rb:
```diff
@@ -0,0 +1,46 @@
+module Middleman
```
Those classes wrap the tests and validate the user input.

dg-ratiodata commented on middleman-core/lib/middleman-core/preview_server/information.rb:
```diff
@@ -0,0 +1,213 @@
+      end
+    end
+
+    class DefaultInformation < BasicInformation
```
This one rejects all unknown combinations of user input. Just in the case.

dg-ratiodata commented on gem_rake_helper.rb:
```diff
@@ -21,7 +21,7 @@ def release_gem(*args)
   exempt_tags << '--tags ~@travishatesme' if ENV['TRAVIS'] == 'true'
-  t.cucumber_opts = "--color #{exempt_tags.join(' ')} --strict --format #{ENV['CUCUMBER_FORMAT'] || 'Fivemat'}"
+  t.cucumber_opts = "--require features --color #{exempt_tags.join(' ')} --strict --format #{ENV['CUCUMBER_FORMAT'] || 'Fivemat'}"
 end
```
This `--require` is necessary otherwise cucumber does not work with `features/cli`-directory. You may want to move some of the `cli_*`-files there too.

maxmeyer commented:
@tdreyno Ok. I rebased and cleaned up the git history a bit. If you prefer to have it squashed into one commit: No problem.

> Fascinating and complicated. Thanks for taking the lead on this!

Ok. I also cleaned up the code. I tried to use the "Single Responsible Principle" where needed. That's why there a quite a few new classes. Maybe we could cleanup the code of the `PreviewServer` a little bit more. But that's a thing for another PR I think.

I removed the offense cleanup of "old" code. That should be done in a separate commit. The default is still `0.0.0.0` and `::` if no commandline option is given. So there's no break on the ui.

Now the `PreviewServer`does not bother the user anymore if the default port is used by some other process. Instead it issues a warning and directly uses a free port to bind the webserver on that.

I tried to write a few feature tests to make it easier to find bugs and document the behaviour of the "new" Preview server. I used some code for this I originally developed for the `aruba`-project. I will notice you if this is not needed anymore because I work on including it in `aruba`'s master branch.

And have a nice holiday! :-)

/mm

tdreyno commented:
Thanks @maxmeyer, I would prefer a single squashed commit if possible.

@bhollis @Arcovion you guys want to take a look at this?

maxmeyer commented:
Ok. Squashed into 2 commits. As one commit changes your test setup and I'm not sure if this is ok for you. The changed functionality is squashed into one single commit.

Arcovion commented:
The bind address option works correctly, I've tested https etc too - this patch doesn't break any existing functionality as far as I can tell (apart from renaming the CLI options). I didn't actually run the test suite on Windows though as Nokogiri doesn't compile for Capybara... but that's unrelated.

The only trouble is the new console output, which looks like this:

    == The Middleman is loading
    == Blog Sources: /posts/:title.html (:prefix + :sources)
    == LiveReload accepting connections from http://10.0.2.15:35729
    == The Middleman is listening on "http://:::4567", "http://0.0.0.0:4567"
    == View your site at "http://localhost:4567", "http://127.0.0.1:4567"
    == Inspect your site configuration at "http://localhost:4567/__middleman", "http://127.0.0.1:4567/__middleman"

Or like this with IPv6 addresses:

    == The Middleman is listening on "http://:::4567", "http://0.0.0.0:4567"
    == View your site at "http://fe80::349d:d97d:12d7:865a%12:4567"
    == Inspect your site configuration at "http://fe80::349d:d97d:12d7:865a%12:4567/__middleman"

I like the seperation where the first line shows the actual addresses that it's listening on, with the viewable URLs below... but the above URL doesn't work. It looked much better before:

    == The Middleman is loading
    == Blog Sources: /posts/:title.html (:prefix + :sources)
    == LiveReload accepting connections from http://10.0.2.15:35729
    == The Middleman is standing watch at http://vagrant-ubuntu-vivid-64:4567/ (http://10.0.2.15:4567/)
    == Inspect your site configuration at http://vagrant-ubuntu-vivid-64:4567/__middleman

Where we had the pretty hostname and a public IP when available.

tdreyno commented:
Yeah, I think the very specific info should go behind the `verbose` flag, while the human readable stuff should be output by default.

Arcovion commented on middleman-core/features/cli/preview_server.feature:
```diff
@@ -0,0 +1,376 @@
+    """
+    And the output should contain:
+    """
+    View your site at "http://::1:4567"
```
IPv6 URLs must [use square brackets](http://superuser.com/a/367785) to work in the browser.

Arcovion commented on middleman-core/features/cli/preview_server.feature:
```diff
@@ -0,0 +1,376 @@
+    """
+    And the output should contain:
+    """
+    View your site at "http://
```
`0.0.0.0` is not a valid IP, it's only routed to `localhost` on Linux - this URL doesn't work on Windows.

Arcovion commented on middleman-core/features/cli/preview_server.feature:
```diff
@@ -0,0 +1,376 @@
+    """
+    And the output should contain:
+    """
+    Inspect your site configuration at "http://127.0.0.1:4567/__middleman"
```
I think this can just be `Inspect your site configuration at /__middleman` - omitting the address.

maxmeyer replied:
Mmmh... But then the user cannot Copy&Paste the address to his/her browser. Would like to leave it like that.

maxmeyer commented:
Sorry, but I broke the build. I need to check this on monday.

maxmeyer commented:
Ok. I found some time over the weekend. The listenening stuff is not meant to be used in a browser. I remove the `http://` before that and renamed the output to make its intent clearer. I also switched from info to debug for that output. The build should be successful now.

maxmeyer commented:
I think I found a solution. Now the host name is added when "0.0.0.0" and "::" and when it is resolved to the ip of a local network interface.

maxmeyer commented:
@tdreyno WDYT, should we go on with merging #1528? Should I include a fix in the PR #1528? For now libv8 is locked through the Gemfile.

maxmeyer commented:
Mmhh.. And would you accept a PR merging the PreviewServer-stuff to master as well?

Arcovion commented:
This is what it looks like for me; missing the hostname:

    == View your site at "http://:4567", "http://192.168.0.3:4567"
    == Inspect your site configuration at "http://:4567/__middleman", "http://192.168.0.3:4567/__middleman"

dg-ratiodata commented:
Not sure why this does not work for you. Could you please add the following to /etc/hosts or the windows equivalent:

```
192.168.0.3 myhost.localdomain myhost
```

and try again?

Arcovion commented:
Yea if I add `192.168.0.3 hostname` to `%SystemRoot%\System32\drivers\etc\hosts` I get:

    == View your site at "http://hostname:4567", "http://192.168.0.3:4567"
    == Inspect your site configuration at "http://hostname:4567/__middleman", "http://192.168.0.3:4567/__middleman"

`Socket.gethostname` or similar would be better of course

dg-ratiodata commented:
Are you behind some kind of consumer (dsl/cable) router with a DNS forwarder? If yes, this cannot work because those devices only forward dns request into the internet but cannot resolve hostnames of your local network. 

The use case described above was the reason for all those checks I added to middleman. Otherwise we would show the hostname, but the user cannot really use the shown hostname to get to the preview server because his/her dns server does not resolve the hostname to the ip of the user's computer.

Arcovion commented:
New version from git looks like this:

    == View your site at "http://192.168.0.3:4567"
    == Inspect your site configuration at "http://192.168.0.3:4567/__middleman"

```
hostname
   Arcovion-PC

nslookup 192.168.0.3
   *** UnKnown can't find 192.168.0.3: No response from server

ipconfig /all
  Host Name . . . . . . . . . . . . : Arcovion-PC
  Link-local IPv6 Address . . . . . : fe80::349d:d97d:12d7:865a%12(Preferred)
  IPv4 Address. . . . . . . . . . . : 192.168.0.3(Preferred)

ruby -r socket -e "puts Socket.gethostname"
  Arcovion-PC
```

I can view the preview server from `http://Arcovion-PC:4567/`,  `http://192.168.0.3:4567/`, `http://localhost:4567/` and `https://...` perfectly fine... it's just not being printed by Middleman.

`ping -a 192.168.0.3` &rarr; `Pinging Arcovion-PC [192.168.0.3] with 32 bytes of data`
` ping -a localhost` &rarr; `Pinging Arcovion-PC [127.0.0.1] with 32 bytes of data`

maxmeyer commented:
Ok. I see, but that seems to be an internal windows thing. The same does not work on Linux. What about OS X? Do they have such magic "things" too? Should we support such things? It's possible but that would operating system specific.

I'm not sure if I like that. If you users want to have this, they could use the /etc/hosts or %SystemRoot%\System32\drivers\etc\hosts and it will work.

dg-ratiodata commented:
Ok. I see, but that seems to be an internal windows thing. The same does not work on Linux. What about OS X? Do they have such magic "things" too? Should we support such things? It's possible but that would operating system specific. Do you know any abstraction (from ruby-core) which supports this? Socket.getnames does not work. I checked this already on windows.

I'm not sure if I like that. If you users want to have this, they could use the /etc/hosts or %SystemRoot%\System32\drivers\etc\hosts and it will work.

dg-ratiodata commented:
Ok. I added some stuff for windows users... Can you try that new version in git?

dg-ratiodata commented:
Mmmh I think it's a different thing, maybe LLMNR is used. Will have a look.

maxmeyer commented:
@Arcovion Can you please check if the new version works for you? Thanks.

Arcovion commented:
Yes, this works!

maxmeyer commented:
Wonderful. I also added a dns stub resolver to the test setup to "secure" this changes. This is why the `Gemfile` inclues `rubydns` now.

maxmeyer commented:
Hope the build is successfull. On a previous version I forgot to check in the Gemfile.

maxmeyer commented:
MDNS is only available on RUBY_VERSION >= 2.1. I added some code for this.

tdreyno commented:
Still failing on ruby <= 2.0, it seems

maxmeyer commented:
Yes, I forgot to disable some checks for ruby < 2.1

maxmeyer commented:
Ok. I upgraded cucumber to 2.x to make scenario skipping possible.

maxmeyer commented:
Thomas, it's all green. :smile: Would you mind to merge?

Cheers,
Dennis

tdreyno commented:
Nice!

I have a quick question about renaming `host`. Is this going to be confusing for Rails users who expect the host/-h option? Is the new feature so different that `host` is no longer an accurate description?

maxmeyer commented:
> Is this going to be confusing for Rails users who expect the host/-h option?

Unfortunately I'm not a regular rails developer. But using `-h` for something other then `--help` bothers me more I must admit. And beside that, `--server-name` is a better description. As it is used in the SSL-certificate as well as Common Name.

Surplus looking at the output of `rails server --help` I can't see a `--host`-option (rails 4.2.3)

```bash
% be rails server -h
Usage: rails server [mongrel, thin etc] [options]
    -p, --port=port                  Runs Rails on the specified port.
                                     Default: 3000
    -b, --binding=IP                 Binds Rails to the specified IP.
                                     Default: localhost
    -c, --config=file                Uses a custom rackup configuration.
    -d, --daemon                     Runs server as a Daemon.
    -u, --debugger                   Enables the debugger.
    -e, --environment=name           Specifies the environment to run this server under (test/development/production).
                                     Default: development
    -P, --pid=pid                    Specifies the PID file.
                                     Default: tmp/pids/server.pid

    -h, --help                       Shows this help message.
```

maxmeyer commented:
BTW: I push 0.7.0 of aruba to rubygems... Would you mind to inform me, if you have any problems? Thanks.

tdreyno commented:
Awesome. Looking for a little thumbs up from @Arcovion and @bhollis

maxmeyer commented:
Ping.

Arcovion commented:
Was this tested on a mac yet? I'm happy with this as long as everyone has tested this works on their machines.

I still prefer the old console output as it was more minimal, this doesn't really need `.local`, quotes, multiple urls for `/__middleman` (which makes the line length > 80 chars). But that's not a big deal.

dg-ratiodata commented:
> (which makes the line length > 80 chars). 

Oh. I understand that. But changing it will require a lot of code changes + the user need to copy a lot of things to build a url she/he can use in a browser. The ".local"-domain is added to be compliant to some RFCs otherwise the application is not required to ask the `LocalLinkResolver` via UDP on 5353.

So if it's ok for you all, I would like to leave the code as it is.

tdreyno commented:
Just got back from holiday. I'll review today.

dg-ratiodata commented:
Hooray... Should I file another PR to include this patchset into v4 as well?

tdreyno commented:
@dg-ratiodata yes please. no rush

maxmeyer commented:
Wonderful. Will file one.

tdreyno commented:
I've got 2 failing tests on Travis from this, @maxmeyer ... I think it may just be around what root-less users can bind to? Fails locally for me on OS X as well.

maxmeyer commented:
Sorry, for my late response. was offline for some days. Will have a look into this next week.

tdreyno commented:
No worries. Thanks @maxmeyer

dg-ratiodata commented:
@tdreyno Can you please try to upgrade cucumber to `2.x`? `#skip_this_scenario` is only available from `cucumber` > 2.0.

tdreyno commented:
I did so, but got the same 2 failing scenarios on Ruby 2.2.1. Should they be getting skipped?

```
      """

Failing Scenarios:
cucumber features/cli/preview_server.feature:118 # Scenario: Start the server with bind address 127.0.0.5
cucumber features/cli/preview_server.feature:362 # Scenario: Start the server when port is blocked by other middleman instance
```

dg-ratiodata commented:
No. Can you do me a favor and add a link to the failing build? I'm going to have a look into this tomorrow morning.

tdreyno commented:
@dg-ratiodata https://travis-ci.org/middleman/middleman/builds/72534677

Pases 2.1+, fails below.
